### PR TITLE
Add an option to control progress reporting

### PR DESCRIPTION
**Pull request by pepeiborra**
The current implementation of progress reporting uses `actionBracket` which has O(logC) complexity in the number of registered Cleanups (https://github.com/ndmitchell/shake/issues/797).

Since we do progress reporting for every File and Rule updated, the worst case complexity is around O(F*R*logC) for all the files in the project. In large projects, this is quite expensive.

Since `otTracedAction` is implemented using `actionBracket` too, I have made it conditional on eventlogs being enabled.

For reviewers understanding: I had to move some types around to avoid cyclic module dependencies.

### CODE REVIEW

**Comment by ndmitchell** [comment]
I think the _O(log n)_ is easily fixable on the Shake side. If it was fixed, would you still want to make this change? When you say "quite expensive" do you have any numbers you can share?

**Comment by pepeiborra** [comment]
> I think the _O(log n)_ is easily fixable on the Shake side. If it was fixed, would you still want to make this change? When you say "quite expensive" do you have any numbers you can share?

Yes, this change still makes sense, to remove the arbitrary exclusion of `GetFileExists` and to make this configurable for users of ghcide as a library. 

The flame graph below is the result of profiling an "edit" experiment over a project with ~40k source files where there is only a FOI which only has 3 transitive dependencies. It shows that ~50% of the time spent in `defineEarlyCutoff` is used by `actionBracket` for the `GetFileExists` and `GetModificationTime` rules which do no work. So 50% of the total overhead induced by `defineEarlyCutoff` is attributable to `actionBracket`

image

**Comment by ndmitchell** [comment]
Ouch, I'll get a constant time actionBracket done today with luck...

File: ghcide/src/Development/IDE/GHC/ExactPrint.hs [hunk 1]
```diff
@@ -67,11 +67,6 @@ import Data.Monoid (All(All))
 #if __GLASGOW_HASKELL__ == 808
 import Control.Arrow
 #endif
-#if __GLASGOW_HASKELL__ > 808
-import Bag (listToBag)
-import ErrUtils (mkErrMsg)
-import Outputable (text, neverQualify)
-#endif
```
Comment (ndmitchell) [inline ghcide/src/Development/IDE/GHC/ExactPrint.hs:74]: Why did this `#if` disappear? Seems unrelated.

File: ghcide/src/Development/IDE/Types/Options.hs [hunk 2]
```diff
@@ -137,8 +130,16 @@ defaultIdeOptions session = IdeOptions
     ,optCheckParents = pure CheckOnSaveAndClose
     ,optHaddockParse = HaddockParse
     ,optCustomDynFlags = id
+    ,optSkipProgress = defaultSkipProgress
     }
 
+defaultSkipProgress :: Typeable a => a -> Bool
+defaultSkipProgress key = case () of
```
Comment (ndmitchell) [inline ghcide/src/Development/IDE/Types/Options.hs:137]: Could we comment as to why these nodes get excluded? The original reason for excluding GetFileExists was that given a module `Foo.hs`, we might have N `GetFileExists` at each include path, at `.lhs`, at `.hs-boot` etc. Whereas every other rule is pretty much only at a single filepath. Is that property true for modtime? It definitely isn't for getfilecontents, but is that about performance?

**Review by ndmitchell — APPROVED** [review]
Looks like a nice simplification and performance saving when we don't have tracing.

File: ghcide/src/Development/IDE/GHC/ExactPrint.hs (hunk 1 above)
Comment (pepeiborra) [reply ghcide/src/Development/IDE/GHC/ExactPrint.hs:74]: doh, again! It's the second time it happens to me, I think this could be a problem with the pre-commit hook, /cc @Ailrun

File: ghcide/src/Development/IDE/Types/Options.hs (hunk 2 above)
Comment (pepeiborra) [reply ghcide/src/Development/IDE/Types/Options.hs:137]: It's performance, but I haven't measured. I'll add a comment or revert to just `GetFileExists`

Comment (ndmitchell) [reply ghcide/src/Development/IDE/Types/Options.hs:137]: I don't think is unreasonable - its a commonly called thing that is mostly irrelevant for performance concerns. I do think its important to say that `GetFileExists` is specially, because its not obvious that is has that weird behaviour, and how if you do include `GetFileExists` your progress is basically meaningless.

File: ghcide/src/Development/IDE/GHC/ExactPrint.hs (hunk 1 above)
Comment (Ailrun) [reply ghcide/src/Development/IDE/GHC/ExactPrint.hs:74]: Could be... It's probably because CPP support for the formatter is not so reliable. Unfortunately, I don't know any formatter works well with CPP...
